plugin/plugins/filesystem/nodes/ipfs.go

```diff
@@ -0,0 +1,222 @@
+
+	if id.Path.Namespace() == nRoot { // metadata should have been initialized by attacher, don't consult CoreAPI
+		return id.Qid, id.metaMask, id.meta, nil
+	}
+
+	if err := coreGetAttr(id.Ctx, &id.meta, req, id.core, id.Path); err != nil {
+		return p9.QID{}, p9.AttrMask{}, p9.Attr{}, err
+	}
+	id.Qid.Type = id.meta.Mode.QIDType()
+
+	metaClone := id.meta
+	metaClone.Filter(req)
+
+	return id.Qid, req, metaClone, nil
+}
+
+func (id *IPFS) Walk(names []string) ([]p9.QID, p9.File, error) {
+	id.Logger.Debugf("Walk names %v", names)
+	id.Logger.Debugf("Walk myself: %s:%v", id.Path, id.Qid)
+
+	if shouldClone(names) {
+		id.Logger.Debugf("Walk cloned")
+		return []p9.QID{id.Qid}, id, nil
+	}
+
+	qids := make([]p9.QID, 0, len(names))
+
+	walkedNode := &IPFS{} // [walk(5)] id represents fid, walkedNode represents newfid
+	*walkedNode = *id
+
+	var err error
+	for _, name := range names {
+		//TODO: timerctx; don't want to hang forever
+		if walkedNode.Path, err = id.core.ResolvePath(id.Ctx, corepath.Join(walkedNode.Path, name)); err != nil {
+			return qids, nil, err
+		}
+
+		ipldNode, err := id.core.Dag().Get(id.Ctx, walkedNode.Path.Cid())
+		if err != nil {
+			return qids, nil, err
+		}
+
+		//TODO: this is too opague; we want core path => qid, dirent isn't necessary
+		dirEnt := &p9.Dirent{}
+		if err = ipldStat(dirEnt, ipldNode); err != nil {
+			return qids, nil, err
+		}
+
+		walkedNode.Qid = dirEnt.QID
+		//
+		qids = append(qids, walkedNode.Qid)
+	}
+
+	id.Logger.Debugf("Walk ret %v, %v", qids, walkedNode)
+	return qids, walkedNode, err
+}
+
+func (id *IPFS) Open(mode p9.OpenFlags) (p9.QID, uint32, error) {
+	id.Logger.Debugf("Open %q", id.Path)
+
+	// set up  handle amenities
+	var handleContext context.Context
+	handleContext, cancel := context.WithCancel(id.Ctx)
+	id.cancel = cancel
+
+	// handle directories
+	if id.meta.Mode.IsDir() {
+		c, err := id.core.Unixfs().Ls(handleContext, id.Path)
+		if err != nil {
+			id.Logger.Errorf("hit\n%q\n%#v\n", id.Path, err)
+			cancel()
+			return id.Qid, 0, err
+		}
+
+		id.directory = &directoryStream{
+			entryChan: c,
+		}
+		return id.Qid, 0, nil
+	}
+
+	// handle files
+	apiNode, err := id.core.Unixfs().Get(handleContext, id.Path)
+	if err != nil {
+		id.Logger.Errorf("hit\n%q\n%#v\n", id.Path, err)
+		cancel()
+		return id.Qid, 0, err
+	}
+
+	var ok bool
+	if id.file, ok = apiNode.(files.File); !ok {
+		cancel()
+		return id.Qid, 0, fmt.Errorf("%q does not appear to be a file: %T", id.Path.String(), apiNode)
+	}
+
+	return id.Qid, ipfsBlockSize, nil
+}
+
+func (id *IPFS) Readdir(offset uint64, count uint32) ([]p9.Dirent, error) {
+	id.Logger.Debugf("Readdir %d %d", offset, count)
+
+	if id.directory == nil {
+		return nil, fmt.Errorf("directory %q is not open for reading", id.Path.String())
+	}
+
+	if offset < 0 {
+		return nil, fmt.Errorf("offset %d can't be negative", offset)
+	}
+
+	if id.directory.eos {
+		if offset == id.directory.cursor-1 {
```

Comment:
Are you sure it's `cursor-1`? The cursor should always point to the _next_ item so the "EOF" case should be `offset == cursor` (unless I'm missing something).

Reply:
We read entries from the `Ls` channel, and when we receive one, we advance the cursor forward.
With this in mind, a full read should leave the cursor pointing at entryCount+1 (0-index), since the final entry will still advance the cursor (beyond the bound).

There may be a better way to handle this, but I'm not sure yet.

Reply:
Given that it's 0 indexed, a full read should leave the cursor at `$entryCount`. If `offset == $entryCount`, the offset is now one past the end of the directory and we're done. If `offset` is `$entryCount-1`, we should return the last item.